### DISCUSSION TITLE
DandDした場合のレベルの遷移

### DISCUSSION THREAD

**Discussion by mochimochinano** [pixiv/vroid-sdk-developers#164] [Q&A]
質問です。

認証レベルでのロードに失敗した場合、DandDでのVRMファイルの読み込みをするように作っています。

認証画面→アバターチェック用レベル→thirdpersonmap　の流れです。

アバターチェック用レベルでDandDをして、正しくVRMファイルが反映したことを確認しても、thirdpersonmapに遷移すると、また初期アバターに戻ってしまいます。

DandDしたファイル情報を保持して、thirdpersonMapのロード時に再適用させる方法はございますでしょうか。

ブループリントベースで製作していますので、BPでご教示いただけますと幸いです。

**Comment by PaperSloth** [comment]
GameInstanceにVrmAssetListObjectをcacheすることで可能かと思います。
明後日辺りにノード含めて回答できればと思いますm(_ _)m

**Reply by mameoPix** [reply]
いくつか手順が必要になります。
1. GameInsntaceを継承したBlueprintを作成
　以下の記事の親クラスがGameInstanceのBlueprintを作成するの欄が参考になるかと思います。
https://zenn.dev/posita33/books/ue5_starter_cpp_and_bp_002/viewer/chap_01-04_bp-gameinstance

2. 1.で作成したGameInstanceをプロジェクトに設定
　BP_GameInstanceというBlueprintを作成して設定しました。
![image](https://github.com/pixiv/vroid-sdk-developers/assets/104663155/cc490cad-7c5e-4c45-b069-ec4f76d0248d)

3. BP_GameInstanceに変数を追加
　CacheAssetListという変数名で、VrmAssetListObjectのオブジェクト参照の変数を追加しました。
![image](https://github.com/pixiv/vroid-sdk-developers/assets/104663155/d5c39b95-4d33-4ada-83f9-b7ebfc993f05)

4. Drag And DropでVRMを更新した際の情報をGameInstanceに保存
　 BP_VRoidMannequinRetargetCharacterのUpdateVRMイベントの後ろに以下のノードを追加。
　更新した際のVRMの情報をGameInstanceに保存しています。
![image](https://github.com/pixiv/vroid-sdk-developers/assets/104663155/eb4f118d-ac80-4e69-9adc-fdd0a198e41e)

5.  4.でキャッシュしたVRMの情報を元にキャラクターのメッシュを更新
　BP_VRoidMannequinRetargetCharacterのBeginPlayの処理でCreatePoseCopyの後に以下のノードを追加。　
![image](https://github.com/pixiv/vroid-sdk-developers/assets/104663155/4aca91ee-01d2-489d-9a06-57da113a95e9)

**Reply by papicochan** [reply]
返信が遅くなり申し訳ありません、無事つくれそうです。ご丁寧にありがとうございました。
